README


This archive contains the distribution Net-Statsd-Client,
version 0.30:

  Send data to StatsD / Graphite

This software is copyright (c) 2013 by Andrew Rodland.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
0.30      2013-09-19 20:38:23 EST5EDT
  * Add 'gauge' and 'set_add' methods to support statsd gauges and sets.
0.22      2013-07-06 14:41:13 EST5EDT
  * Depend on Etsy::StatsD 1.001. This fixes a bug which would sometimes
    cause corrupted partial metrics to be sent (thanks Ricardo Signes and
    Steve Sanbeg).
  * Fix tests to pass with Etsy::StatsD 1.001.
0.21      2013-06-13 17:48:49 America/New_York
  * Allow setting a callback that will be used to emit a warning when a
    timer is destroyed unexpectedly. Idea courtesy of rjbs.
0.20      2013-06-12 17:15:47 America/New_York
  * Use Moo. This makes the code a little more manageable and has a
    negligible impact on speed.
  * Don't run failure-prone "live" tests on install or CPANtesters, add a
    set of non-live tests.
  * Add the ability to re-target the metric of a timer between starting and
    finishing it.
0.11      2013-06-03 15:28:30 America/New_York
  * Make timers report their times in ms like they should
0.10      2013-05-19 21:33:44 EST5EDT
  * First public release.
0.03
  * Internal Shutterstock release
  * Extracted from Shutterstock::Statsd


lib/Net/Statsd/Client.pm
package Net::Statsd::Client;
use Moo;
use Sub::Quote;

# ABSTRACT: Send data to StatsD / Graphite
our $VERSION = '0.30'; # VERSION
our $AUTHORITY = 'cpan:ARODLAND'; # AUTHORITY

use Etsy::StatsD 1.001;
use Net::Statsd::Client::Timer;

has 'prefix' => (
  is => 'ro',
  default => quote_sub q{''},
);

has 'sample_rate' => (
  is => 'ro',
  default => quote_sub q{1},
);

has 'host' => (
  is => 'ro',
  default => quote_sub q{'localhost'},
);

has 'port' => (
  is => 'ro',
  default => quote_sub q{8125},
);

has 'statsd' => (
  is => 'rw',
);

has 'warning_callback' => (
  is => 'rw',
);

sub BUILD {
  my ($self) = @_;
  $self->statsd(
    Etsy::StatsD->new($self->host, $self->port)
  );
}

sub _send {
  my ($self, $data, $sample_rate) = @_;

  $self->{statsd}->send(
    { "$self->{prefix}$data->{metric}" => "$data->{value}|$data->{type}" },
    ( defined($data->{sample_rate}) ? $data->{sample_rate} : $self->{sample_rate} ),
  );
}
    
sub increment {
  # ($self, $metric, [$sample_rate])
  $_[0]->_send({ metric => $_[1], value => 1, type => "c", sample_rate => $_[2] });
}

sub decrement {
  # ($self, $metric, [$sample_rate])
  $_[0]->_send({ metric => $_[1], value => -1, type => "c", sample_rate => $_[2] });
}

sub update {
  # ($self, $metric, $value, [$sample_rate])
  $_[0]->_send({ metric => $_[1], value => $_[2], type => "c", sample_rate => $_[3] });
}

sub timing_ms {
  # ($self, $metric, $time, [$sample_rate])
  $_[0]->_send({ metric => $_[1], value => $_[2], type => "ms", sample_rate => $_[3] });
}

sub gauge {
  # ($self, $metric, $value, [$sample_rate])
  $_[0]->_send({ metric => $_[1], value => $_[2], type => "g", sample_rate => $_[3] });
}

sub set_add {
  # ($self, $metric, $value, [$sample_rate])
  $_[0]->_send({ metric => $_[1], value => $_[2], type => "s", sample_rate => $_[3] });
}

sub timer {
  my ($self, $metric, $sample_rate) = @_;

  return Net::Statsd::Client::Timer->new(
    statsd => $self,
    metric => $metric,
    sample_rate => $sample_rate,
    warning_callback => $self->warning_callback,
  );
}

1;



=pod

=head1 NAME

Net::Statsd::Client - Send data to StatsD / Graphite

=head1 VERSION

version 0.30

=head1 SYNOPSIS

    use Net::Statsd::Client
    my $stats = Net::Statsd::Client->new(prefix => "service.frobnitzer.");
    $stats->increment("requests"); # service.frobnitzer.requests++ in graphite

    my $timer = $stats->timer("request_duration");
    # ... do something expensive ...
    $timer->finish;

=head1 ATTRIBUTES

=head2 host

B<Optional:> The hostname of the StatsD server to connect to. Defaults to
localhost.

=head2 port

B<Optional:> The port number to connect to. Defaults to 8125.

=head2 prefix

B<Optional:> A prefix to be added to all metric names logged throught his
object.

=head2 sample_rate

B<Optional:> A value between 0 and 1, determines what fraction of events
will actually be sent to the server. This sets the default sample rate,
which can be overridden on a case-by-case basis when sending an event (for
instance, you might choose to send errors at a 100% sample rate, but other
events at 1%).

=head2 warning_callback

B<Optional:> A function that will be called with a message if a C<timer>
is destroyed unexpectedly (see L<Net::Statsd::Timer>). If this is not set
the builtin C<warn> will be used.

=head1 METHODS

=head2 $stats->increment($metric, [$sample_rate])

Increment the named counter metric.

=head2 $stats->decrement($metric, [$sample_rate])

Decrement the named counter metric.

=head2 $stats->update($metric, $count, [$sample_rate])

Add C<$count> to the value of the named counter metric.

=head2 $stats->timing_ms($metric, $time, [$sample_rate])

Record an event of duration C<$time> milliseconds for the named timing metric.

=head2 $stats->timer($metric, [$sample_rate])

Returns a L<Net::Statsd::Client::Timer> object for the named timing metric.
The timer begins when you call this method, and ends when you call C<finish>
on the timer.

=head2 $stats->gauge($metric, $value, [$sample_rate])

Send a value for the named gauge metric. Instead of adding up like counters
or producing a large number of quantiles like timings, gauges simply take
the last value sent in any time period, and don't require scaling.

=head2 $statsd->set_add($metric, $value, [$sample_rate])

Add a value to the named set metric. Sets count the number of *unique*
values they see in each time period, letting you estimate, for example, the
number of users using a site at a time by adding their userids to a set each
time they load a page.

=head1 AUTHOR

Andrew Rodland <arodland@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Andrew Rodland.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__



lib/Net/Statsd/Client/Timer.pm
package Net::Statsd::Client::Timer;
use Moo;
use Sub::Quote;

# ABSTRACT: Measure event timings and send them to StatsD
our $VERSION = '0.30'; # VERSION
our $AUTHORITY = 'cpan:ARODLAND'; # AUTHORITY

use Time::HiRes qw(gettimeofday tv_interval);

has 'statsd' => (
  is => 'ro',
  required => 1,
);

has '_pending' => (
  is => 'rw',
  default => quote_sub q{1},
);

has ['metric', 'start', '_file', '_line', 'warning_callback'] => (
  is => 'rw',
);

sub BUILD {
  my ($self) = @_;
  my (undef, $file, $line) = caller(1);

  $self->start([gettimeofday]);
  $self->_file($file);
  $self->_line($line);
}

sub finish {
  my ($self) = @_;
  my $duration = tv_interval($self->{start});
  $self->{statsd}->timing_ms(
    $self->{metric},
    $duration * 1000,
    $self->{sample_rate},
  );
  delete $self->{_pending};
}

sub cancel {
  my ($self) = @_;
  delete $self->{_pending};
}

sub emit_warning {
  my $self = shift;
  if (defined $self->warning_callback) {
    $self->warning_callback->(@_);
  } else {
    warn(@_);
  }
}

sub DEMOLISH {
  my ($self) = @_;
  if ($self->{_pending}) {
    my $metric = $self->{metric};
    $metric = $self->{statsd}{prefix} . $metric if $self->{statsd} && $self->{statsd}{prefix};
    $self->emit_warning("Unfinished timer for stat $metric (created at $self->{_file} line $self->{_line})");
  }
}

1;



=pod

=head1 NAME

Net::Statsd::Client::Timer - Measure event timings and send them to StatsD

=head1 VERSION

version 0.30

=head1 SYNOPSIS

    use Net::Statsd::Client;
    my $stats = Net::Statsd::Client->new(prefix => "service.frobnitzer.");

    my $timer = $stats->timer("request_duration");
    # ... do something expensive ...
    $timer->finish;

=head1 METHODS

=head2 Net::Statsd::Client::Timer->new(...)

To build a timer object, call L<Net::Statsd::Client>'s C<timer> method,
instead of calling this constructor directly.

A timer has an associated statsd object, metric name, and sample rate, and
begins counting as soon as it's constructed.

=head2 $timer->finish

Stop timing, and send the elapsed time to the server.

=head2 $timer->cancel

Stop timing, but do not send the elapsed time to the server. A timer that
goes out of scope without having C<finish> or C<cancel> called on it will
generate a warning, since this probably points to bugs and lost timing
information.

=head2 $timer->metric($new)

Change the metric name of a timer on the fly. This is useful if you don't
know what kind of event you're timing until it's finished. Harebrained
example:

    my $timer = $statsd->timer("item.fetch");
    my $item = $cache->get("blah");
    if ($item) {
        $timer->metric("item.fetch_from_cache");
    } else {
        $item = get_it_the_long_way();
    }
    $timer->finish;

=head1 AUTHOR

Andrew Rodland <arodland@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Andrew Rodland.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__



Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Send data to StatsD / Graphite",
  "AUTHOR" => "Andrew Rodland <arodland\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Net-Statsd-Client",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Net::Statsd::Client",
  "PREREQ_PM" => {
    "Etsy::StatsD" => "1.001",
    "Moo" => 0,
    "Sub::Quote" => 0,
    "Time::HiRes" => 0
  },
  "TEST_REQUIRES" => {
    "Exporter" => 0,
    "FindBin" => 0,
    "IO::Socket" => 0,
    "Test::More" => 0,
    "lib" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.30",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



